File: java-devildyw/src/main/java/com/dyw/leetcode/simple/tree/L637二叉树的层平均值.java
package com.dyw.leetcode.simple.tree;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * 给定一个非空二叉树的根节点 root , 以数组的形式返回每一层节点的平均值。与实际答案相差 10-5 以内的答案可以被接受。
 *
 * @author Devil
 * @since 2022-09-21-10:41
 */
@SuppressWarnings("all")
public class L637二叉树的层平均值 {

    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> result = new ArrayList<>();
        Deque<TreeNode> deque = new LinkedList<>();
        if (root==null){
            return result;
        }
        deque.offer(root);

        while (!deque.isEmpty()){
            int len = deque.size();
            int n = len;
            double sum = 0.0;
            while (len-->0){
                TreeNode node = deque.poll();
                if (node.left!=null){
                    deque.offer(node.left);
                }
                if (node.right!=null){
                    deque.offer(node.right);
                }
                sum += node.val;
            }
            result.add(sum/n);
        }
        return result;
    }
}

File: golang-devildyw/simple/L637二叉树的层平均值.go
package main

func main() {}

func averageOfLevels(root *TreeNode) []float64 {
	if root == nil {
		return []float64{}
	}

	result := []float64{}
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		length := len(queue)
		sum := 0
		for i := 0; i < length; i++ {
			first := queue[0]
			queue = queue[1:]
			sum += first.Val
			if first.Left != nil {
				queue = append(queue, first.Left)
			}
			if first.Right != nil {
				queue = append(queue, first.Right)
			}
		}
		result = append(result, float64(sum)/float64(length))
	}
	return result
}
